pkg/promise/promise.go

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package promise
+
+import (
+	"sync"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// A promise for a future value.
+type Promise[T any] interface {
+	// Await blocks until the value is resolved.
+	Await() T
+}
+
+// New creates a new promise for value T.
+// Returns function to resolve the promise, and the promise itself.
+func New[T any]() (func(T), Promise[T]) {
+	p := &promise[T]{}
+	p.cond = sync.NewCond(p)
+	return p.resolve, p
+}
+
+// Map transforms the value of a promise with the provided function.
+func Map[A, B any](p Promise[A], transform func(A) B) Promise[B] {
+	resolve, p2 := New[B]()
+	go func() {
+		resolve(transform(p.Await()))
+	}()
+	return p2
+}
+
+type promise[T any] struct {
+	lock.Mutex
+	cond     *sync.Cond
+	resolved bool
+	value    T
+}
+
+func (p *promise[T]) resolve(value T) {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.resolved {
+		return
+	}
+	p.resolved = true
+	p.value = value
+	p.cond.Broadcast()
+}
+
+// Await blocks until the promise has been resolved.
+func (p *promise[T]) Await() T {
```

Comment:
Does it make sense to pass in a context here to allow a caller to cancel the await? Most Promise APIs I have seen in Go allow this.

It would require you to use a different internal mechanism than CondVars though.

Reply:
I did think about it, but for now I wanted to keep this to bare minimum (no timeouts, no failed promises) and only used in initialization. Though maybe it might make sense to call it something else as the API right now does not match with what one usually thinks of promises. 

Meh screw it, I'll just add in the missing features.

Reply:
@gandro PTAL

Reply:
Ah nil ctx is bad form. Will remove the optionality and fix the gofmt fail after back from lunch…